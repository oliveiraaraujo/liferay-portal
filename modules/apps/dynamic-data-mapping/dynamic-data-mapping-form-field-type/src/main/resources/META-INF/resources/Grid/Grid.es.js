/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {ClayInput, ClayRadio} from '@clayui/form';
import ClayTable from '@clayui/table';
import React, {useState} from 'react';

import {FieldBaseProxy} from '../FieldBase/ReactFieldBase.es';
import getConnectedReactComponentAdapter from '../util/ReactComponentAdapter.es';
import {connectStore} from '../util/connectStore.es';
import templates from './GridAdapter.soy';

const TableHead = ({columns}) => (
	<ClayTable.Head>
		<ClayTable.Row>
			<ClayTable.Cell headingCell />
			{columns.map((column, colIndex) => {
				if (column.value) {
					return (
						<ClayTable.Cell
							headingCell
							key={`column-${column.value}-${colIndex}`}
						>
							{column.label}
						</ClayTable.Cell>
					);
				}
			})}
		</ClayTable.Row>
	</ClayTable.Head>
);

const TableBodyColumns = ({
	columns,
	disabled,
	onBlur,
	onChange,
	onFocus,
	row,
	rowIndex,
	value,
}) =>
	columns.map((column, colIndex) => {
		if (column.value) {
			return (
				<ClayTable.Cell key={`cell-${column.value}-${colIndex}`}>
					<ClayRadio
						aria-label={`grid_${rowIndex}_${colIndex}`}
						checked={column.value === value[row.value]}
						className="form-builder-grid-field"
						disabled={disabled}
						name={row.value}
						onBlur={onBlur}
						onChange={onChange}
						onFocus={onFocus}
						value={column.value}
					/>
				</ClayTable.Cell>
			);
		}
	});

const Grid = ({
	columns = [{label: 'col1', value: 'fieldId'}],
	disabled,
	name,
	onBlur = () => {},
	onChange = () => {},
	onFocus = () => {},
	rows = [{label: 'row', value: 'jehf'}],
	value,
	...otherProps
}) => (
	<div className="table-responsive" {...otherProps}>
		{!disabled &&
			rows.map((row, rowIndex) => {
				const inputValue = value[row.value]
					? `${row.value};${value[row.value]}`
					: '';

				return (
					<ClayInput
						aria-label="grid_hidden"
						key={`row-${row.value}-${rowIndex}`}
						name={name}
						type="hidden"
						value={inputValue}
					/>
				);
			})}

		<ClayTable striped>
			<TableHead columns={columns} />

			<ClayTable.Body>
				{rows.map((row, rowIndex) => {
					if (row.value) {
						return (
							<ClayTable.Row
								key={`row-${row.value}-${rowIndex}`}
								name={row.value}
							>
								<ClayTable.Cell>{row.label}</ClayTable.Cell>

								<TableBodyColumns
									columns={columns}
									disabled={disabled}
									onBlur={onBlur}
									onChange={onChange}
									onFocus={onFocus}
									row={row}
									rowIndex={rowIndex}
									value={value}
								/>
							</ClayTable.Row>
						);
					}

					return null;
				})}
			</ClayTable.Body>
		</ClayTable>
	</div>
);

const GridProxy = connectStore(
	({columns, emit, name, readOnly, rows, value = {}, ...otherProps}) => {
		const [state, setState] = useState(value);

		return (
			<FieldBaseProxy {...otherProps}>
				<Grid
					columns={columns}
					disabled={readOnly}
					name={name}
					onBlur={event =>
						emit('fieldBlurred', event, event.target.value)
					}
					onChange={event => {
						const {target} = event;
						const value = {
							[target.name]: target.value,
						};

						const newState = {...state, ...value};

						setState(newState);

						emit('fieldEdited', event, newState);
					}}
					onFocus={event =>
						emit('fieldFocused', event, event.target.value)
					}
					rows={rows}
					value={state}
				/>
			</FieldBaseProxy>
		);
	}
);

const ReactGridAdapter = getConnectedReactComponentAdapter(
	GridProxy,
	templates
);

export {ReactGridAdapter};

export default ReactGridAdapter;
