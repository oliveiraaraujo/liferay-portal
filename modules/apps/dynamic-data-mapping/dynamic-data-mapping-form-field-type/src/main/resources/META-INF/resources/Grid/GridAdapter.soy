{namespace GridAdapter}

/**
 * GridAdapter
 */
{template .render}
	<div key="ddm-field-type-grid"></div>
{/template}