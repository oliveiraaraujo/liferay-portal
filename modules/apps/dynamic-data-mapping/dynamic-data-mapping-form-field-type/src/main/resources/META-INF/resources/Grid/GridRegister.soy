{namespace GridRegister}

{deltemplate PageRenderer.RegisterFieldType variant="'grid'"}
	{call GridAdapter.render data="all" /}
{/deltemplate}