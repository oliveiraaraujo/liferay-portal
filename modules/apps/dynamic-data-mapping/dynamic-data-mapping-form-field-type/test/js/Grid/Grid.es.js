/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {fireEvent} from '@testing-library/react';

import Grid from '../../../src/main/resources/META-INF/resources/Grid/Grid.es';
import withContextMock from '../__mocks__/withContextMock.es';

let component;
const spritemap = 'icons.svg';

const GridWithContextMock = withContextMock(Grid);

describe('Grid', () => {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('renders columns', () => {
		component = new Grid({
			columns: [
				{
					label: 'col1',
					value: 'fieldId',
				},
				{
					label: 'col2',
					value: 'fieldId',
				},
			],
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('renders no columns when columns comes empty', () => {
		component = new Grid({
			columns: [],
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('is not edidable', () => {
		component = new Grid({
			readOnly: false,
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('has a tip', () => {
		component = new Grid({
			spritemap,
			tip: 'Type something',
		});

		expect(component).toMatchSnapshot();
	});

	it('has an id', () => {
		component = new Grid({
			id: 'ID',
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('has a label', () => {
		component = new Grid({
			label: 'label',
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('is not required', () => {
		component = new Grid({
			required: false,
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('renders rows', () => {
		component = new Grid({
			rows: [
				{
					label: 'row1',
					value: 'fieldId',
				},
				{
					label: 'row2',
					value: 'fieldId',
				},
			],
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('renders no rows when row comes empty', () => {
		component = new Grid({
			rows: [],
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('renders Label if showLabel is true', () => {
		component = new Grid({
			label: 'text',
			showLabel: true,
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('emits a fieldBlurred event when blurring the radio input', done => {
		const handleFieldBlurred = data => {
			expect(data).toEqual(
				expect.objectContaining({
					value: 'colFieldId1'
				})
			);
			done();
		};
		
		const events = {fieldBlurred: handleFieldBlurred};

		component = new GridWithContextMock({
			columns: [
				{
					label: 'col1',
					value: 'colFieldId1',
				},
				{
					label: 'col2',
					value: 'colFieldId2',
				},
			],
			events,
			readOnly: false,
			rows: [
				{
					label: 'row1',
					value: 'rowFieldId1',
				},
				{
					label: 'row2',
					value: 'rowFieldId2',
				},
			],
			spritemap,
		});

		const radioInputElement = component.element.querySelector(
			'input[value][type="radio"][name="rowFieldId1"]:not([value="colFieldId2"])'
		);

		fireEvent.blur(radioInputElement);
	});

	it('emits a fieldEdited event when changing the state of radio input', done => {
		const handleFieldEdited = data => {
			expect(data).toEqual(
				expect.objectContaining({
					fieldInstance: expect.any(Object),
					originalEvent: expect.any(Object),
					value: {
						rowFieldId1: 'colFieldId1'
					}
				})
			);
			done();
		};
		
		const events = {fieldEdited: handleFieldEdited};

		component = new GridWithContextMock({
			columns: [
				{
					label: 'col1',
					value: 'colFieldId1',
				},
				{
					label: 'col2',
					value: 'colFieldId2',
				},
			],
			events,
			readOnly: false,
			rows: [
				{
					label: 'row1',
					value: 'rowFieldId1',
				},
				{
					label: 'row2',
					value: 'rowFieldId2',
				},
			],
			spritemap,
		});

		const radioInputElement = component.element.querySelector(
			'input[value][type="radio"][name="rowFieldId1"]:not([value="colFieldId2"])'
		);

		fireEvent.click(radioInputElement);
	});

	it('emits a fieldFocused event when focusing a radio input', done => {
		const handleFieldFocused = data => {
			expect(data).toEqual(
				expect.objectContaining({
					fieldInstance: expect.any(Object),
					originalEvent: expect.any(Object),
					value: 'colFieldId1'
				})
			);
			done();
		};
		
		const events = {fieldFocused: handleFieldFocused};

		component = new GridWithContextMock({
			columns: [
				{
					label: 'col1',
					value: 'colFieldId1',
				},
				{
					label: 'col2',
					value: 'colFieldId2',
				},
			],
			events,
			readOnly: false,
			rows: [
				{
					label: 'row1',
					value: 'rowFieldId1',
				},
				{
					label: 'row2',
					value: 'rowFieldId2',
				},
			],
			spritemap,
		});

		const radioInputElement = component.element.querySelector(
			'input[value][type="radio"][name="rowFieldId1"]:not([value="colFieldId2"])'
		);

		fireEvent.focus(radioInputElement);
	});
});
