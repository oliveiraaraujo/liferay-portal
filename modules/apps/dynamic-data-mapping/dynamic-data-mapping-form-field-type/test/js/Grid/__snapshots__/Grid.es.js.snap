// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid has a label 1`] = `
<div class="react-component-adapter">
  <div class="form-group">
    <p class="ddm-label">label</p>
    <div class="table-responsive"><input aria-label="grid_hidden" type="hidden" class="form-control" value="">
      <div class="table-responsive">
        <table class="table table-autofit show-quick-actions-on-hover table-hover table-list table-striped">
          <thead>
            <tr class="">
              <th class=""></th>
              <th class="">col1</th>
            </tr>
          </thead>
          <tbody>
            <tr name="jehf" class="">
              <td class="">row</td>
              <td class="">
                <div class="custom-control custom-radio"><label><input aria-label="grid_0_0" name="jehf" class="custom-control-input form-builder-grid-field" role="radio" type="radio" value="fieldId"><span class="custom-control-label"></span></label></div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid has a tip 1`] = `
<div class="react-component-adapter">
  <div class="form-group">
    <div class="table-responsive"><input aria-label="grid_hidden" type="hidden" class="form-control" value="">
      <div class="table-responsive">
        <table class="table table-autofit show-quick-actions-on-hover table-hover table-list table-striped">
          <thead>
            <tr class="">
              <th class=""></th>
              <th class="">col1</th>
            </tr>
          </thead>
          <tbody>
            <tr name="jehf" class="">
              <td class="">row</td>
              <td class="">
                <div class="custom-control custom-radio"><label><input aria-label="grid_0_0" name="jehf" class="custom-control-input form-builder-grid-field" role="radio" type="radio" value="fieldId"><span class="custom-control-label"></span></label></div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div><span class="form-text">Type something</span>
  </div>
</div>
`;

exports[`Grid has an id 1`] = `
<div class="react-component-adapter">
  <div class="form-group">
    <div class="table-responsive"><input aria-label="grid_hidden" type="hidden" class="form-control" value="">
      <div class="table-responsive">
        <table class="table table-autofit show-quick-actions-on-hover table-hover table-list table-striped">
          <thead>
            <tr class="">
              <th class=""></th>
              <th class="">col1</th>
            </tr>
          </thead>
          <tbody>
            <tr name="jehf" class="">
              <td class="">row</td>
              <td class="">
                <div class="custom-control custom-radio"><label><input aria-label="grid_0_0" name="jehf" class="custom-control-input form-builder-grid-field" role="radio" type="radio" value="fieldId"><span class="custom-control-label"></span></label></div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid is not edidable 1`] = `
<div class="react-component-adapter">
  <div class="form-group">
    <div class="table-responsive"><input aria-label="grid_hidden" type="hidden" class="form-control" value="">
      <div class="table-responsive">
        <table class="table table-autofit show-quick-actions-on-hover table-hover table-list table-striped">
          <thead>
            <tr class="">
              <th class=""></th>
              <th class="">col1</th>
            </tr>
          </thead>
          <tbody>
            <tr name="jehf" class="">
              <td class="">row</td>
              <td class="">
                <div class="custom-control custom-radio"><label><input aria-label="grid_0_0" name="jehf" class="custom-control-input form-builder-grid-field" role="radio" type="radio" value="fieldId"><span class="custom-control-label"></span></label></div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid is not required 1`] = `
<div class="react-component-adapter">
  <div class="form-group">
    <div class="table-responsive"><input aria-label="grid_hidden" type="hidden" class="form-control" value="">
      <div class="table-responsive">
        <table class="table table-autofit show-quick-actions-on-hover table-hover table-list table-striped">
          <thead>
            <tr class="">
              <th class=""></th>
              <th class="">col1</th>
            </tr>
          </thead>
          <tbody>
            <tr name="jehf" class="">
              <td class="">row</td>
              <td class="">
                <div class="custom-control custom-radio"><label><input aria-label="grid_0_0" name="jehf" class="custom-control-input form-builder-grid-field" role="radio" type="radio" value="fieldId"><span class="custom-control-label"></span></label></div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid renders Label if showLabel is true 1`] = `
<div class="react-component-adapter">
  <div class="form-group">
    <p class="ddm-label">text</p>
    <div class="table-responsive"><input aria-label="grid_hidden" type="hidden" class="form-control" value="">
      <div class="table-responsive">
        <table class="table table-autofit show-quick-actions-on-hover table-hover table-list table-striped">
          <thead>
            <tr class="">
              <th class=""></th>
              <th class="">col1</th>
            </tr>
          </thead>
          <tbody>
            <tr name="jehf" class="">
              <td class="">row</td>
              <td class="">
                <div class="custom-control custom-radio"><label><input aria-label="grid_0_0" name="jehf" class="custom-control-input form-builder-grid-field" role="radio" type="radio" value="fieldId"><span class="custom-control-label"></span></label></div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid renders columns 1`] = `
<div class="react-component-adapter">
  <div class="form-group">
    <div class="table-responsive"><input aria-label="grid_hidden" type="hidden" class="form-control" value="">
      <div class="table-responsive">
        <table class="table table-autofit show-quick-actions-on-hover table-hover table-list table-striped">
          <thead>
            <tr class="">
              <th class=""></th>
              <th class="">col1</th>
              <th class="">col2</th>
            </tr>
          </thead>
          <tbody>
            <tr name="jehf" class="">
              <td class="">row</td>
              <td class="">
                <div class="custom-control custom-radio"><label><input aria-label="grid_0_0" name="jehf" class="custom-control-input form-builder-grid-field" role="radio" type="radio" value="fieldId"><span class="custom-control-label"></span></label></div>
              </td>
              <td class="">
                <div class="custom-control custom-radio"><label><input aria-label="grid_0_1" name="jehf" class="custom-control-input form-builder-grid-field" role="radio" type="radio" value="fieldId"><span class="custom-control-label"></span></label></div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid renders no columns when columns comes empty 1`] = `
<div class="react-component-adapter">
  <div class="form-group">
    <div class="table-responsive"><input aria-label="grid_hidden" type="hidden" class="form-control" value="">
      <div class="table-responsive">
        <table class="table table-autofit show-quick-actions-on-hover table-hover table-list table-striped">
          <thead>
            <tr class="">
              <th class=""></th>
            </tr>
          </thead>
          <tbody>
            <tr name="jehf" class="">
              <td class="">row</td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid renders no rows when row comes empty 1`] = `
<div class="react-component-adapter">
  <div class="form-group">
    <div class="table-responsive">
      <div class="table-responsive">
        <table class="table table-autofit show-quick-actions-on-hover table-hover table-list table-striped">
          <thead>
            <tr class="">
              <th class=""></th>
              <th class="">col1</th>
            </tr>
          </thead>
          <tbody></tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid renders rows 1`] = `
<div class="react-component-adapter">
  <div class="form-group">
    <div class="table-responsive"><input aria-label="grid_hidden" type="hidden" class="form-control" value=""><input aria-label="grid_hidden" type="hidden" class="form-control" value="">
      <div class="table-responsive">
        <table class="table table-autofit show-quick-actions-on-hover table-hover table-list table-striped">
          <thead>
            <tr class="">
              <th class=""></th>
              <th class="">col1</th>
            </tr>
          </thead>
          <tbody>
            <tr name="fieldId" class="">
              <td class="">row1</td>
              <td class="">
                <div class="custom-control custom-radio"><label><input aria-label="grid_0_0" name="fieldId" class="custom-control-input form-builder-grid-field" role="radio" type="radio" value="fieldId"><span class="custom-control-label"></span></label></div>
              </td>
            </tr>
            <tr name="fieldId" class="">
              <td class="">row2</td>
              <td class="">
                <div class="custom-control custom-radio"><label><input aria-label="grid_1_0" name="fieldId" class="custom-control-input form-builder-grid-field" role="radio" type="radio" value="fieldId"><span class="custom-control-label"></span></label></div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
